21 2024 analysis

crispy-wonton opened the pull request:
Fixes #21 
---

# Description

Updated analysis to use latest MCS and EPC data (Q1 2024).

Summary of changes:
- I have updated the relevant supplementary datasets where required. Other datasets are the same as in the previous analysis*.
- Update config and getters so that they get the relevant supplementary data from the appropriate subdirectory
- Update getters processing steps for new datasets where required
- Update README with data sourcing

*Datasets that have note been changed/updated:
- Rurality data is the same because 2011 rurality data is still the latest available dataset
- tenure and dwelling data have not been updated because they both use 2021 census data which I believe are the latest datasets available. 


# Instructions for Reviewer

To test the code, make sure you have a local dir of `ASF_data` containing the deduplicated and preprocessed EPC data for Q1 2024. Then run:
`python -i asf_welsh_energy_consultation/analysis/produce_plots_and_stats.py --local_data_dir /YOUR/LOCAL/DIR`

Please pay special attention to ...
- please check you agree that the appropriate supplementary datasets have been updated
- check correct sources have been linked in README and attribution statements are correct

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [ ] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
emancipate functions
Functions were not able to run independently outside of script because they depended on global variables which originated from argparser in
get_data.py. This meant that most, if not all functions, could not run independently because most depend on one of the getter functions. This
meant the functions could only be used if running the final script as intended. They could not be used independently in notebooks for example.
This made testing and use of individual functions difficult. This commit emancipates the functions so that they can be used independently.
- move argparser from get_data.py to produce_plots_and_stats.py and only run argparser if __name__ == __main__
- remove global variables defined in get_data.py and process_data.py
- add appropriate args to functions in process_data.py and get_data.py so that they function without global variables and argparser
- update docstrings with new args and improve some docstrings for clarity
- update check_local_epc function in get_data.py to only require one `epc_version` input from user

crispy-wonton committed:
apply changes requested in PR review - clarity of variable names and docstrings

crispy-wonton committed:
update base config to identify supplementary data by edition

crispy-wonton committed:
update get_data.py getters to process 2024 datasets and to allow for auto-retrieval of appropriate supplementary datasets based on input year& month of analysis

crispy-wonton committed:
Merge branch 'dev' into 21_2024_analysis

crispy-wonton committed:
update README with data sources and attributions

crispy-wonton committed:
correct source data date in README

crispy-wonton committed:
add missing brackets to README

sofiapinto commented on README.md:
```diff
@@ -74,29 +74,25 @@ outputs/
-- rurality.ods - 2011 Rural Urban Classification for small area geographies, see [here](https://www.ons.gov.uk/methodology/geography/geographicalproducts/ruralurbanclassifications)
-- postcodes ; postcode_to_output_area.csv ; tenure.csv - ONS data with unknown source date
-- dwellings data not historically used in this analysis
+Versions of data used for previous analysis are listed below.
```
Absolutely love this table - so organised and easy to read 👯

sofiapinto commented on README.md:
```diff
@@ -74,29 +74,25 @@ outputs/
-- dwellings data not historically used in this analysis
+Versions of data used for previous analysis are listed below.
+
+| Analysis      | EPC                                                                | MCS<sup>1</sup> | Postcodes<sup>2</sup>                                                                              | Postcode to OA<sup>3</sup>                                                                                                                                  | Off gas postcodes<sup>4</sup>       ...
```
I think it's also worth have a `<sup>` for EPC

sofiapinto commented on README.md:
```diff
@@ -74,29 +74,25 @@ outputs/
+| April 2023    | 2022 Q4 complete (preprocessed)                                    | 230315          | ONS postcode directory. Date unknown.                                                              | ONS data. Date unknown.                                                                                                                                     | [2022](https://www.xoserve.com/help ...
+
+# Data attributions
+
```
```suggestion
1. Energy Performance Certificate (EPC) Register data. England and Wales data available [here](https://epc.opendatacommunities.org/) and Scotland data available [here](https://statistics.gov.scot/resource?uri=http%3A%2F%2Fstatistics.gov.scot%2Fdata%2Fdomestic-energy-performance-certificates). Data has undergone cleaning and processing before analysis - the code available in this [Github repository](https://github.com/nestauk/asf_core_data)
```

sofiapinto commented on README.md:
```diff
@@ -74,29 +74,25 @@ outputs/
+
+# Data attributions
+
+1. MCS Installations Database. Date represents internal processing date.
```
Some rewording + additions. Feel free to reword further obv :)
```suggestion
2. Microgeneration Certification Scheme (MCS) Installations Database (MID), including data about heat pump installations. The numbers under the MCS column in the table represent internal processing dates and he analyses make use of `mcs_installations_YYMMDD.csv` and `mcs_installations_epc_full_YYMMDD.csv`.
```

sofiapinto commented on README.md:
```diff
@@ -74,29 +74,25 @@ outputs/
+| ------------- | ------------------------------------------------------------------ | --------------- | -------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------- ...
+| November 2024 | 2024 Q1 complete (preprocessed, and preprocessed and deduplicated) | 241113          | [February 2024](https://geoportal.statistics.gov.uk/datasets/e14b1475ecf74b58804cf667b6740706)     | [February 2024](https://geoportal.statistics.gov.uk/datasets/ons::postcode-to-oa-2021-to-lsoa-to-msoa-to-lad-february-2024-best-fit-lookup-in-the-uk/about) | [September 2024](https://www.xoserv ...
+| October 2023  | 2023 Q2 complete (preprocessed, and preprocessed and deduplicated) | 231009          | [August 2023](https://geoportal.statistics.gov.uk/datasets/487a5ba62c8b4da08f01eb3c08e304f6/about) | [May 2022](https://geoportal.statistics.gov.uk/datasets/e7824b1475604212a2325cd373946235/about)                                                             | [2022](https://www.xoserve.com/help ...
+| April 2023    | 2022 Q4 complete (preprocessed)                                    | 230315          | ONS postcode directory. Date unknown.                                                              | ONS data. Date unknown.                                                                                                                                     | [2022](https://www.xoserve.com/help ...
```
```suggestion
| April 2023    | 2022 Q4 complete (preprocessed)                                    | 230315          | ONS postcode directory. Date unknown.                                                              | ONS data. Date unknown.                                                                                                                                     | [2022](https://www.xoserve.com/help-centre/supply-points-metering/supply-point-administration-spa/)           | [2011](https://www.gov.uk/government/statistics/2011-rural-urban-classification-lookup-tables-for-all-geographies) | Not used                                                                                                                                       | ONS census data. Date unknown.                                                       |
```

sofiapinto commented on README.md:
```diff
@@ -74,29 +74,25 @@ outputs/
+| November 2024 | 2024 Q1 complete (preprocessed, and preprocessed and deduplicated) | 241113          | [February 2024](https://geoportal.statistics.gov.uk/datasets/e14b1475ecf74b58804cf667b6740706)     | [February 2024](https://geoportal.statistics.gov.uk/datasets/ons::postcode-to-oa-2021-to-lsoa-to-msoa-to-lad-february-2024-best-fit-lookup-in-the-uk/about) | [September 2024](https://www.xoserv ...
+| October 2023  | 2023 Q2 complete (preprocessed, and preprocessed and deduplicated) | 231009          | [August 2023](https://geoportal.statistics.gov.uk/datasets/487a5ba62c8b4da08f01eb3c08e304f6/about) | [May 2022](https://geoportal.statistics.gov.uk/datasets/e7824b1475604212a2325cd373946235/about)                                                             | [2022](https://www.xoserve.com/help ...
+| April 2023    | 2022 Q4 complete (preprocessed)                                    | 230315          | ONS postcode directory. Date unknown.                                                              | ONS data. Date unknown.                                                                                                                                     | [2022](https://www.xoserve.com/help ...
+
```
```suggestion

_To note:_ "Analysis" relates to when the analysis was ran. The remaining columns have information about the versions of data used in each analysis.

```

sofiapinto commented on asf_welsh_energy_consultation/config/base.yaml:
```diff
@@ -3,9 +3,19 @@ plots:
-  tenure_data: "tenure.csv"
-  dwelling_data: "dwellings_2021.xlsx"
+  data_202411:
+    postcode_dir: "ONSPD_FEB_2024_UK.csv"
```
it's weird to call this `postcode_dir`, but I understand it was a directory in the previous analysis. maybe we could call it `postcode_data` instead?

crispy-wonton replied:
It's called `postcode_dir` because the dataset is called the Postcode Directory

sofiapinto commented on asf_welsh_energy_consultation/config/base.yaml:
```diff
@@ -3,9 +3,19 @@ plots:
+    rurality_data: "rurality.ods"
+    tenure_data: "tenure.csv"
+    dwelling_data: "dwellings_2021.xlsx"
+  data_202304:
```
what happenned to data_202310? there's no information here.

crispy-wonton replied:
This has now been corrected

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -57,37 +57,46 @@ def get_countries(input_data_path):
-        )
+
+    # From 202411 analysis, we use the full postcode directory rather than the partitioned files
+    if data_edition > "data_202310":
```
just so that the documentation matches the code, for easier read
```suggestion
    if data_edition >= "data_202411":
```

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -244,20 +258,28 @@ def get_rurality_by_oa(input_data_path):
-    oa = oa[["pcd7", "oa11cd"]].rename(
-        columns={"pcd7": "postcode", "oa11cd": "oa_code"}
+    # Historical data requires a different encoding
+    if data_edition > "data_202310":
```
```suggestion
    if data_edition >= "data_202411":
```

sofiapinto commented on asf_welsh_energy_consultation/getters/get_data.py:
```diff
@@ -244,20 +258,28 @@ def get_rurality_by_oa(input_data_path):
-        columns={"pcd7": "postcode", "oa11cd": "oa_code"}
+    # Historical data requires a different encoding
+    if data_edition > "data_202310":
+        oa = pd.read_csv(PROJECT_DIR / oa_path)
```
Just a note that sometimes we're using `/` and other times we're using `os.path.join()`

crispy-wonton replied:
Yes, this is due to the legacy code. I didnt think it was worth going through and updating everything to `os.path.join()` but have used it where I added new lines as it's more robust.

sofiapinto replied:
sounds good to me!

sofiapinto reviewed (commented):
Hello Roisin :) @crispy-wonton 

Changes to the code:
- `README.md`: I looked at the changes to the README including the data versions and attributions. I really like the new table. It makes it really easy to visualise the information. Have left some comments with suggestions for improvements, but feel free to only accept if it makes sense to you. The updated versions of the datasets look good.
- `base.yaml`: I like the new way of organising the information about the different analyses. I think we're missing the information about `data_202310` though.
- `get_data.py`: The changes look good. I proposed minor changes.

Running the analysis - I did the following steps:
- `make install`
- `conda activate asf_welsh_energy_consultation`
- `make inputs-pull`
- Downloaded `2024_Q1_complete/EPC_GB_preprocessed_and_deduplicated.csv` and 2024_Q1_complete/EPC_GB_preprocessed.csv` from S3 and moved them to my `outputs/EPC/preprocessed_data/2024_Q1_complete` within my local ASF data folder.
- Ran `python -i asf_welsh_energy_consultation/analysis/produce_plots_and_stats.py --local_data_dir /YOUR/LOCAL/DIR` with my local dir.

It all ran okay :)

When you provide the "Estimated percentage of properties with a heat pump" to Marine don't forget to reinforce that the number is based on MCS installations only, and hence not representing all properties with a heat pump in Wales.

A note on version:
In the README it says "You can specify which batch of EPC data to download and MCS data to load from S3 by passing the --epc_batch and --mcs_batch arguments, both default to downloading/loading the newest data from S3, respectively." Maybe we could just reinforce that people should set the epc_batch and mcs_batch (even if it's the newest) just to make sure we're using the right versions?

crispy-wonton committed:
update README documentation with changes requested in code review

crispy-wonton committed:
update base yaml with 202304 source data and correct 202310 label

crispy-wonton committed:
Update get_data.py for code readability

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

crispy-wonton committed:
Update get_data.py for code readability

Co-authored-by: Ana Sofia Pinto <70645959+sofiapinto@users.noreply.github.com>

crispy-wonton committed:
update README with information on MCS and EPC batches

crispy-wonton commented:
> When you provide the "Estimated percentage of properties with a heat pump" to Marine don't forget to reinforce that the number is based on MCS installations only, and hence not representing all properties with a heat pump in Wales.

Good point, will do.

> A note on version: In the README it says "You can specify which batch of EPC data to download and MCS data to load from S3 by passing the --epc_batch and --mcs_batch arguments, both default to downloading/loading the newest data from S3, respectively." Maybe we could just reinforce that people should set the epc_batch and mcs_batch (even if it's the newest) just to make sure we're using the right versions?

I have updated the README.

Thank you for your review @sofiapinto and for testing out the code by running it. I have addressed or replied to all your comments. Please let me know if you're happy to merge.

sofiapinto commented:
wow queen, you are speedy! happy for you to merge 🥇
